src/pcp.rs

```diff
@@ -801,33 +780,66 @@ mod tests {
             Ok(res)
         }
 
-        fn valid_gadget_calls(&self) -> Vec<usize> {
-            vec![2, 1]
+        fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+            vec![
+                Box::new(Mul::new(2)),
+                Box::new(PolyEval::new(poly_range_check(2, 5), 1)),
+            ]
         }
 
-        fn joint_rand_len(&self) -> usize {
-            1
+        fn as_slice(&self) -> &[F] {
+            &self.data
         }
 
-        fn prove_rand_len(&self) -> usize {
-            3
+        fn param(&self) -> TestValueParam {
+            TestValueParam()
         }
 
-        fn query_rand_len(&self) -> usize {
+        fn into_output(self) -> Vec<F> {
+            self.data
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    struct TestValueParam();
```

Comment:
nit: You can define an empty struct and still derive Clone and Debug: 
```suggestion
    struct TestValueParam;
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=667df556946b5fab68f8d6dfa59bb588

Reply:
Thanks for the tip :)